File: tests/hash_check_test.py
#!/usr/bin/env python3
"""
The status page's hash check over HTTP (api/hash_check.php):
    python tests/hash_check_test.py         (needs the local server: php -S 127.0.0.1:8089, and the smoke fixtures)

What only HTTP can show: the gate (a visitor without status.hash_check is refused, a member is not),
the rate limit (a setting, per address, 429 past it), one hash spelled three ways giving one answer,
and what a bad input gets. The lookup itself is tests/hash_check_test.php.
"""
import base64
import http.cookiejar
import json
import os
import re
import subprocess
import sys
import urllib.error
import urllib.request

BASE = os.environ.get("SMOKE_BASE", "http://127.0.0.1:8089/")
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
BOOT = ("$root = '" + ROOT.replace("\\", "/") + "';"
        "require_once $root . '/config/app.php'; require_once $root . '/config/database.php';"
        "require_once $root . '/includes/settings.php'; require_once $root . '/includes/functions.php';"
        "require_once $root . '/includes/schema.php'; require_once $root . '/includes/whitelist.php';"
        "require_once $root . '/includes/mail.php'; require_once $root . '/includes/users.php';"
        "$db = getDb(); $cfg = getSettings($db); ensureSchema($db, $cfg);")

fails = 0
n = 0


def check(name, cond, info=""):
    global fails, n
    n += 1
    print(("PASS " if cond else "FAIL ") + name + ("" if cond or not info else "  -> " + str(info)[:300]))
    if not cond:
        fails += 1


def php(code):
    p = subprocess.run(["php", "-d", "error_reporting=E_ALL & ~E_DEPRECATED", "-r", BOOT + code],
                       capture_output=True, text=True, cwd=ROOT, encoding="utf-8", errors="replace")
    if p.returncode != 0:
        print("php failed: " + (p.stderr or p.stdout)[:400])
    return p.stdout


def clear_throttles():
    for f in ("login_attempts.json", "rate_limits.json", "rate_limits.json.lock"):
        try:
            os.remove(os.path.join(ROOT, "config", f))
        except OSError:
            pass


class Client:
    def __init__(self):
        self.jar = http.cookiejar.CookieJar()
        self.opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(self.jar))
        self.csrf = None

    def page(self, action):
        with self.opener.open(BASE + "?action=" + action, timeout=30) as r:
            html = r.read().decode("utf-8", "replace")
        m = re.search(r'name="csrf_token" value="([^"]+)"', html)
        if m:
            self.csrf = m.group(1)
        return r.status, html

    def api(self, endpoint, method="GET", body=None):
        url = BASE + "api.php?endpoint=" + endpoint
        data = json.dumps(body).encode() if body is not None else None
        h = {"Accept": "application/json"}
        if data is not None:
            h["Content-Type"] = "application/json"
        req = urllib.request.Request(url, data=data, method=method, headers=h)
        try:
            with self.opener.open(req, timeout=30) as r:
                return r.status, json.loads(r.read().decode() or "{}")
        except urllib.error.HTTPError as e:
            try:
                return e.code, json.loads(e.read().decode() or "{}")
            except Exception:
                return e.code, {}


HASH = "1c1e" * 10
SEEN = "1c2d" * 10          # the index's own hash, with words on it: the two gated sections
B32 = base64.b32encode(bytes.fromhex(HASH)).decode()
MAGNET = "magnet:?xt=urn:btih:" + HASH.upper() + "&dn=Hash+check+fixture&tr=udp%3A%2F%2Ftracker.example.org%3A6969%2Fannounce"

# Its own member account. The stateful PHP suites truncate `users` on their way out, and this runs
# after them in the battery, so the smoke fixtures' smokeuser is not something this test may rely on.
USER, PASS = "hcuser", "SmokePass123!"
clear_throttles()
was_users = php("echo $cfg['users_enabled'] ?? '0';").strip()
php("$db->prepare('DELETE FROM users WHERE username = ?')->execute(['" + USER + "']);"
    "userCreate($db, $cfg, '" + USER + "', 'hc@example.org', '" + PASS + "', '127.0.0.1');"
    "$db->exec(\"UPDATE users SET email_verified = 1 WHERE username = '" + USER + "'\");")
# The grant is put on the member group HERE, not assumed from the migration: the stateful PHP
# suites rewrite the group's permissions before this runs (tests/users_test.php), and the migration's
# one-time marker is already in place, so what the migration did is tests/hash_check_test.php's
# business. This test is about the gate: with the grant, an answer; without it, a refusal.
GRANT = ("$db->exec(" + chr(34) + "UPDATE user_groups SET permissions = JSON_MERGE_PATCH(permissions, "
         + chr(34) + " . $db->quote(json_encode(['status.hash_check' => %s])) . " + chr(34)
         + ") WHERE slug = 'member'" + chr(34) + ");")
php(GRANT % "true")


# The sections of the answer are gated by the permissions that publish the same facts on the Info
# panel (includes/hashcheck.php: hashCheckGates), so each of those is put on the group and taken off
# it here, one at a time. Over HTTP, because userEffectivePermissions() memoises per process and a
# revocation inside one would not be seen.
def grant(perms, slug="member"):
    pairs = ", ".join("'%s' => %s" % (k, "true" if v else "false") for k, v in perms.items())
    php("$db->exec(\"UPDATE user_groups SET permissions = JSON_MERGE_PATCH(permissions, \" . $db->quote(json_encode([" + pairs + "])) . \") WHERE slug = '" + slug + "'\");")


was_limit = php("echo $cfg['rate_limit_hash_check'] ?? '120';").strip()
was_index = {k: php("echo $cfg['" + k + "'] ?? '';").strip()
             for k in ("index_enabled", "index_search_enabled", "index_search_include_whitelist")}
member_before = php("echo $db->query(\"SELECT permissions FROM user_groups WHERE slug = 'member'\")->fetchColumn();").strip()
php("setSetting($db, 'users_enabled', '1'); setSetting($db, 'rate_limit_hash_check', '120');"
    "setSetting($db, 'index_enabled', '1'); setSetting($db, 'index_search_enabled', '1');"
    "setSetting($db, 'index_search_include_whitelist', '1');"
    "$db->prepare('DELETE FROM whitelist WHERE info_hash = ?')->execute(['" + HASH + "']);"
    "$db->prepare(\"INSERT INTO whitelist (info_hash, name, source, created_at, banned, probe_status) VALUES (?, 'Hash check fixture', 'admin', '2026-09-02 09:00:00', 0, 'passed')\")->execute(['" + HASH + "']);"
    "foreach (['index_hashes', 'hash_content'] as $t) $db->prepare(\"DELETE FROM `$t` WHERE info_hash = ?\")->execute(['" + SEEN + "']);"
    "$db->prepare(\"INSERT INTO index_hashes (info_hash, name, first_seen, last_seen, seen_count, last_seeders, last_leechers, meta_status, files_count)"
    " VALUES (?, 'Seen over HTTP', '2026-08-01 08:00:00', '2026-09-10 20:30:00', 7, 11, 2, 'done', 4)\")->execute(['" + SEEN + "']);"
    "$db->prepare(\"INSERT INTO hash_content (info_hash, description, description_format, content_status)"
    " VALUES (?, 'Words about it.', 'bbcode', 'approved')\")->execute(['" + SEEN + "']);")
grant({"index.view": True, "whitelist.view": True, "content.view": True})

try:
    # ── the gate ──────────────────────────────────────────────────────────
    guest = Client()
    guest.page("status")
    s, j = guest.api("hash_check&hash=" + HASH)
    check("a visitor without the permission is refused", s == 403 and "error" in j, (s, j))
    s, html = guest.page("status")
    check("… and the form is not on the page for them", 'id="hc-form"' not in html)

    member = Client()
    member.page("login")
    s, j = member.api("user_login", "POST", {"csrf_token": member.csrf, "login": USER, "password": PASS})
    check("the member signs in", s == 200 and j.get("success"), (s, j))
    s, j = member.api("user_me")
    check("the member carries status.hash_check", "status.hash_check" in (j.get("permissions") or []), (s, j))
    s, html = member.page("status")
    check("the form is on the page for them", 'id="hc-form"' in html and 'id="hc-result"' in html)

    # ── the answer, and one hash spelled three ways ───────────────────────
    s, a = member.api("hash_check&hash=" + HASH)
    check("a registered hash answers 200 with its state", s == 200 and a.get("success") and (a.get("registered") or {}).get("state") == "live", (s, a))
    check("… known, not banned, never seen", a.get("known") is True and a.get("banned") is None and a.get("seen") is None, a)
    s, b = member.api("hash_check&hash=" + B32)
    s2, c = member.api("hash_check&hash=" + urllib.request.quote(MAGNET, safe=""))
    check("the base32 spelling and the magnet give the same hash", b.get("hash") == HASH and c.get("hash") == HASH, (b.get("hash"), c.get("hash")))
    check("… and the same answer", b.get("registered") == a.get("registered") and c.get("registered") == a.get("registered"))
    s, u = member.api("hash_check&hash=" + "1c1f" * 10)
    check("a hash nobody has met is answered too, as unknown", s == 200 and u.get("known") is False and u.get("registered") is None, (s, u))

    # ── one permission buys the question, not every answer ────────────────
    s, full = member.api("hash_check&hash=" + SEEN)
    check("with all three catalogue permissions the answer is whole",
          s == 200 and (full.get("seen") or {}).get("times") == 7 and (full.get("content") or {}).get("status") == "approved"
          and full.get("meta", {}).get("name") == "Seen over HTTP" and full.get("withheld") == [], (s, full))

    grant({"index.view": False})
    s, j = member.api("hash_check&hash=" + SEEN)
    check("without index.view there is no swarm section", s == 200 and j.get("seen") is None and "seen" in (j.get("withheld") or []), (s, j))
    check("… and no name, metadata or file counts out of the index row either",
          j.get("meta", {}).get("name") is None and j.get("meta", {}).get("status") == "none"
          and j.get("files", {}).get("fetched") == 0 and j.get("files", {}).get("total") is None, j)
    check("… but the hash is still known, and a ban would still be answered", j.get("known") is True and j.get("banned") is None, j)
    grant({"index.view": True})

    grant({"whitelist.view": False})
    s, j = member.api("hash_check&hash=" + HASH)
    check("without whitelist.view there is no registered section", s == 200 and j.get("registered") is None
          and "registered" in (j.get("withheld") or []) and j.get("known") is True, (s, j))
    grant({"whitelist.view": True})

    grant({"content.view": False})
    s, j = member.api("hash_check&hash=" + SEEN)
    check("without content.view there is no content section, and the rest is untouched",
          s == 200 and j.get("content") is None and "content" in (j.get("withheld") or [])
          and (j.get("seen") or {}).get("times") == 7, (s, j))
    grant({"content.view": True})

    php("setSetting($db, 'index_search_enabled', '0');")
    s, j = member.api("hash_check&hash=" + SEEN)
    check("the search switched off closes the swarm section for everybody", s == 200 and j.get("seen") is None, (s, j))
    php("setSetting($db, 'index_search_enabled', '1');")
    php("setSetting($db, 'index_search_include_whitelist', '0');")
    s, j = member.api("hash_check&hash=" + HASH)
    check("a search configured without whitelist rows closes the registered section", s == 200 and j.get("registered") is None, (s, j))
    php("setSetting($db, 'index_search_include_whitelist', '1');")

    # ── what a bad input gets ─────────────────────────────────────────────
    s, e = member.api("hash_check&hash=")
    check("an empty input is a 400 with a sentence", s == 400 and e.get("error"), (s, e))
    s, e = member.api("hash_check&hash=not-a-hash")
    check("garbage is a 400 with a sentence", s == 400 and e.get("error"), (s, e))
    s, e = member.api("hash_check&hash=" + urllib.request.quote("magnet:?dn=no-hash-here", safe=""))
    check("a magnet without a btih is a 400", s == 400 and e.get("error"), (s, e))

    # ── the gate is the grant, nothing else ───────────────────────────────
    php(GRANT % "false")
    s, e = member.api("hash_check&hash=" + HASH)
    check("with the grant taken away the same member is refused", s == 403, (s, e))
    s, html = member.page("status")
    check("… and the form leaves their page too", 'id="hc-form"' not in html)
    php(GRANT % "true")

    # ── the limit ─────────────────────────────────────────────────────────
    php("setSetting($db, 'rate_limit_hash_check', '3');")
    clear_throttles()
    codes = [member.api("hash_check&hash=" + HASH)[0] for _ in range(4)]
    check("three checks pass and the fourth is 429 with the limit at 3", codes == [200, 200, 200, 429], codes)
    php("setSetting($db, 'rate_limit_hash_check', '0');")
    clear_throttles()
    codes = [member.api("hash_check&hash=" + HASH)[0] for _ in range(5)]
    check("0 means no limit", codes == [200] * 5, codes)
finally:
    php("setSetting($db, 'rate_limit_hash_check', '" + (was_limit or "120") + "'); setSetting($db, 'users_enabled', '" + (was_users or "1") + "');"
        + "".join("setSetting($db, '%s', '%s');" % (k, v) for k, v in was_index.items() if v != "")
        + "$db->prepare('DELETE FROM whitelist WHERE info_hash = ?')->execute(['" + HASH + "']);"
        "foreach (['index_hashes', 'hash_content'] as $t) $db->prepare(\"DELETE FROM `$t` WHERE info_hash = ?\")->execute(['" + SEEN + "']);"
        "$db->prepare('DELETE FROM users WHERE username = ?')->execute(['" + USER + "']);"
        "$db->prepare(\"UPDATE user_groups SET permissions = ? WHERE slug = 'member'\")->execute([" + json.dumps(member_before) + "]);")
    clear_throttles()

print("\n%d checks, %d failed" % (n, fails))
sys.exit(1 if fails else 0)

File: tests/hash_check_test.php
<?php
/**
 * The status page's "what does this tracker know about this hash?" (includes/hashcheck.php):
 *   php tests/hash_check_test.php
 *
 * Runs against the local test database (deploy/local_bootstrap.php first). Six fixture hashes, one
 * per state the answer can be in, inserted and removed here; and the registry side of the feature:
 * the permission exists, the member preset carries it, the migration granted it, the limit has a
 * default. The endpoint's gate and rate limit are tests/hash_check_test.py (they need HTTP).
 */
if (PHP_SAPI !== 'cli') { http_response_code(404); exit; }
$root = dirname(__DIR__);
require_once $root . '/config/app.php';
require_once $root . '/config/database.php';
require_once $root . '/includes/settings.php';
require_once $root . '/includes/functions.php';
require_once $root . '/includes/schema.php';
require_once $root . '/includes/whitelist.php';
require_once $root . '/includes/users.php';
require_once $root . '/includes/hashcheck.php';

$fails = 0; $n = 0;
function check(string $name, bool $ok, string $info = ''): void {
    global $fails, $n; $n++;
    echo ($ok ? 'PASS ' : 'FAIL ') . $name . ($ok || $info === '' ? '' : '  -> ' . $info) . "\n";
    if (!$ok) $fails++;
}

$db = getDb();
$cfg = getSettings($db);
ensureSchema($db, $cfg);
$cfg = getSettings($db, true);

// ── the registry side ────────────────────────────────────────────────────────
check('the permission exists in the registry', isset(userPermissionList()['status.hash_check']));
check('the member preset carries it', in_array('status.hash_check', userGroupPresets()['member']['perms'], true));
check('the migration recorded its grant', isset($cfg['schema_grant_v57_hash_check']));
// The grant is a one-time migration and this is a shared database other suites edit (users_test
// resets the system groups to their seed for its own checks). Ask the migration itself rather than
// the leftovers: forget the marker, run the data migrations again — idempotent by design — and read
// what they wrote.
$db->exec("DELETE FROM settings WHERE `key` = 'schema_grant_v57_hash_check'");
trackerSchemaDataMigrations($db, getSettings($db, true));
$cfg = getSettings($db, true);
check('… and records it again when asked to', isset($cfg['schema_grant_v57_hash_check']));
$st = $db->prepare("SELECT permissions FROM user_groups WHERE slug = 'member'");
$st->execute();
$memberPerms = json_decode((string)$st->fetchColumn(), true) ?: [];
check('… and the member group has it on this (migrated) database', ($memberPerms['status.hash_check'] ?? null) === true, json_encode($memberPerms));
$st = $db->prepare("SELECT permissions FROM user_groups WHERE slug = 'guest'");
$st->execute();
$guestPerms = json_decode((string)$st->fetchColumn(), true) ?: [];
check('… and the guest group does not', empty($guestPerms['status.hash_check']), json_encode($guestPerms));
check('the limit ships with a default', (trackerSchemaDefaultSettings()['rate_limit_hash_check'] ?? null) === '120');
check('with accounts switched off the legacy fallback opens it (like content.*), not closes it', userLegacyDefault('status.hash_check') === true);

// ── fixtures: one hash per state ─────────────────────────────────────────────
$H = fn(string $tag) => str_repeat($tag, 10);   // 4 hex chars × 10 = 40
$live = $H('1c01'); $review = $H('1c02'); $probing = $H('1c03'); $rejected = $H('1c04');
$failed = $H('1c05'); $wlBanned = $H('1c06'); $seenOnly = $H('1c07'); $bannedOnly = $H('1c08');
$both = $H('1c09'); $unknown = $H('1c0a');
$all = [$live, $review, $probing, $rejected, $failed, $wlBanned, $seenOnly, $bannedOnly, $both, $unknown];
$clean = function () use ($db, $all) {
    $ph = implode(',', array_fill(0, count($all), '?'));
    foreach (['whitelist', 'index_hashes', 'index_files', 'banned_hashes'] as $t) {
        $db->prepare("DELETE FROM `$t` WHERE info_hash IN ($ph)")->execute($all);
    }
};
$clean();
$wl = $db->prepare("INSERT INTO whitelist (info_hash, name, source, created_at, banned, review_status, probe_status, content_status, meta_status, files_count)
                    VALUES (?, ?, 'admin', '2026-09-01 10:00:00', ?, ?, ?, ?, ?, ?)");
$wl->execute([$live,     'Live fixture',     0, 'none',     'passed',  'approved', 'done',   3]);
$wl->execute([$review,   'Review fixture',   0, 'pending',  'none',    'none',     'none',   null]);
$wl->execute([$probing,  'Probing fixture',  0, 'none',     'probing', 'none',     'none',   null]);
$wl->execute([$rejected, 'Rejected fixture', 0, 'rejected', 'none',    'none',     'none',   null]);
$wl->execute([$failed,   'Failed fixture',   0, 'none',     'failed',  'none',     'failed', null]);
$wl->execute([$wlBanned, 'Banned fixture',   1, 'none',     'passed',  'none',     'done',   1]);
$wl->execute([$both,     'Both fixture',     0, 'none',     'passed',  'none',     'pending', 9000]);
$ix = $db->prepare("INSERT INTO index_hashes (info_hash, name, first_seen, last_seen, seen_count, last_seeders, last_leechers, meta_status, files_count)
                    VALUES (?, ?, '2026-08-01 08:00:00', '2026-09-10 20:30:00', ?, ?, ?, ?, ?)");
$ix->execute([$seenOnly, 'Seen fixture', 17, 120, 4, 'done', 2]);
$ix->execute([$both,     'Both fixture (index)', 3, 5, 1, 'done', 9000]);
$if = $db->prepare("INSERT INTO index_files (info_hash, path, size) VALUES (?, ?, ?)");
$if->execute([$seenOnly, 'a/one.bin', 10]); $if->execute([$seenOnly, 'a/two.bin', 20]);
foreach (range(1, 5) as $k) $if->execute([$both, "part-$k.bin", 100]);
$db->prepare("INSERT INTO banned_hashes (info_hash, reason, source, created_at) VALUES (?, 'fixture', 'admin', '2026-09-05 12:00:00')")->execute([$bannedOnly]);
$db->prepare("INSERT INTO banned_hashes (info_hash, reason, source, created_at) VALUES (?, 'fixture', 'admin', '2026-09-06 12:00:00')")->execute([$wlBanned]);

try {
    $r = hashCheckLookup($db, $cfg, strtoupper($live));
    check('the hash comes back lower-case whatever was given', $r['hash'] === $live, $r['hash']);
    check('a live whitelist row: registered, state live, since its date', ($r['registered']['state'] ?? '') === 'live' && ($r['registered']['since'] ?? '') === '2026-09-01 10:00:00', json_encode($r['registered']));
    check('… carries the published description flag and the name', ($r['registered']['content'] ?? '') === 'approved' && ($r['registered']['name'] ?? '') === 'Live fixture');
    check('… is not banned, never seen, known', $r['banned'] === null && $r['seen'] === null && $r['known'] === true, json_encode($r));
    check('… metadata from the whitelist row, files: 0 stored of 3', $r['meta']['status'] === 'done' && $r['meta']['source'] === 'whitelist'
          && $r['files']['fetched'] === 0 && $r['files']['total'] === 3, json_encode($r['files']));

    check('a row waiting for review says so', hashCheckLookup($db, $cfg, $review)['registered']['state'] === 'review');
    check('a row waiting for its first peer says so', hashCheckLookup($db, $cfg, $probing)['registered']['state'] === 'probing');
    check('a rejected submission says so', hashCheckLookup($db, $cfg, $rejected)['registered']['state'] === 'rejected');
    check('a row whose probe found nothing says so', hashCheckLookup($db, $cfg, $failed)['registered']['state'] === 'failed');
    $r = hashCheckLookup($db, $cfg, $wlBanned);
    check('a banned whitelist row is "banned" whatever its probe says, and the ban list agrees',
          $r['registered']['state'] === 'banned' && $r['banned'] !== null && $r['banned']['since'] === '2026-09-06 12:00:00', json_encode($r));

    $r = hashCheckLookup($db, $cfg, $seenOnly);
    check('a hash only the index knows: not registered, seen', $r['registered'] === null && $r['seen'] !== null && $r['known'] === true);
    check('… with when, how often and how big', $r['seen']['first'] === '2026-08-01 08:00:00' && $r['seen']['last'] === '2026-09-10 20:30:00'
          && $r['seen']['times'] === 17 && $r['seen']['seeders'] === 120 && $r['seen']['leechers'] === 4, json_encode($r['seen']));
    check('… metadata from the index, and both file numbers', $r['meta']['source'] === 'index' && $r['meta']['name'] === 'Seen fixture'
          && $r['files']['fetched'] === 2 && $r['files']['total'] === 2, json_encode($r));

    $r = hashCheckLookup($db, $cfg, $bannedOnly);
    check('a hash only the ban list knows: banned, nothing else, known', $r['banned'] !== null && $r['registered'] === null && $r['seen'] === null && $r['known']);

    $r = hashCheckLookup($db, $cfg, $both);
    check('a hash in both tables: the index speaks for the metadata', $r['meta']['source'] === 'index' && $r['meta']['name'] === 'Both fixture (index)', json_encode($r['meta']));
    check('… and the stored file count is the real one, beside the torrent\'s own', $r['files']['fetched'] === 5 && $r['files']['total'] === 9000, json_encode($r['files']));

    $r = hashCheckLookup($db, $cfg, $unknown);
    check('a hash nobody has met: unknown in every column', !$r['known'] && $r['banned'] === null && $r['registered'] === null && $r['seen'] === null
          && $r['meta']['status'] === 'none' && $r['files']['fetched'] === 0 && $r['files']['total'] === null, json_encode($r));
    check('the answer names the mode, so the page can say "blacklisted" where that is the word', in_array($r['mode'], ['whitelist', 'blacklist'], true), $r['mode']);
} finally {
    $clean();
}

echo "\n$n checks, $fails failed\n";
exit($fails ? 1 : 0);
